File: horizontal-picker/src/main/java/com/github/jhonnyx2012/horizontalpicker/HorizontalPickerRecyclerView.java
package com.github.jhonnyx2012.horizontalpicker;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.LinearSmoothScroller;
import android.support.v7.widget.LinearSnapHelper;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.View;

import org.threeten.bp.LocalDate;
import org.threeten.bp.temporal.ChronoUnit;

/**
 * Created by Jhonny Barrios on 22/02/2017.
 */

public class HorizontalPickerRecyclerView extends RecyclerView implements OnItemClickedListener, View.OnClickListener {

  private final static int OFFSET = 3;
  private HorizontalPickerAdapter adapter;
  private LinearLayoutManager layoutManager;
  private float itemWidth;
  private HorizontalPickerListener listener;
  private boolean scrollToSelectedPosition;

  public HorizontalPickerRecyclerView(Context context) {
    super(context);
  }

  public HorizontalPickerRecyclerView(Context context, @Nullable AttributeSet attrs) {
    super(context, attrs);
  }

  public HorizontalPickerRecyclerView(Context context, @Nullable AttributeSet attrs, int defStyle) {
    super(context, attrs, defStyle);
  }

  public void init(Context context,
                   final int daysToPlus,
                   final int initialOffset,
                   final boolean scrollToSelectedPosition,
                   final int mBackgroundColor,
                   final int mDateSelectedColor,
                   final int mDateSelectedTextColor,
                   final int mTodayDateTextColor,
                   final int mTodayDateBackgroundColor,
                   final int mDayOfWeekTextColor,
                   final int mUnselectedDayTextColor) {
    this.scrollToSelectedPosition = scrollToSelectedPosition;
    layoutManager = new LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false);
    setLayoutManager(layoutManager);
    post(new Runnable() {
      @Override
      public void run() {
        itemWidth = getMeasuredWidth() / 7;
        adapter = new HorizontalPickerAdapter((int) itemWidth, HorizontalPickerRecyclerView.this, getContext(), daysToPlus, initialOffset, mBackgroundColor, mDateSelectedColor, mDateSelectedTextColor, mTodayDateTextColor,
            mTodayDateBackgroundColor,
            mDayOfWeekTextColor,
            mUnselectedDayTextColor,
            OFFSET);
        setAdapter(adapter);
        LinearSnapHelper snapHelper = new LinearSnapHelper();
        snapHelper.attachToRecyclerView(HorizontalPickerRecyclerView.this);
        removeOnScrollListener(onScrollListener);
        addOnScrollListener(onScrollListener);
      }
    });
  }

  private OnScrollListener onScrollListener = new OnScrollListener() {
    @Override
    public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
      super.onScrollStateChanged(recyclerView, newState);
      switch (newState) {
        case RecyclerView.SCROLL_STATE_IDLE:
          listener.onStopDraggingPicker();
          int position = (int) ((computeHorizontalScrollOffset() / itemWidth) + 3.5);
          if (position != -1 && position != adapter.getSelectedPosition()) {
            selectItem(position);
          }
          break;
        case SCROLL_STATE_DRAGGING:
          listener.onDraggingPicker();
          break;
      }
    }

    @Override
    public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
      super.onScrolled(recyclerView, dx, dy);
    }
  };

  private void selectItem(int position) {
    if (position == NO_POSITION) {
      return;
    }
    if (position != adapter.getSelectedPosition()) {
      int unselectPosition = adapter.getSelectedPosition();
      adapter.setSelectedPosition(position);
      adapter.notifyItemChanged(unselectPosition);
      adapter.notifyItemChanged(position);
      listener.onDateSelected(adapter.getItem(position));
    }
  }

  public void setListener(HorizontalPickerListener listener) {
    this.listener = listener;
  }

  @Override
  public void onClickView(View v, int adapterPosition) {
    if (adapterPosition != adapter.getSelectedPosition()
        && adapter.getItemViewType(adapterPosition) != R.layout.item_placeholder) {
      selectItem(adapterPosition);
      if (scrollToSelectedPosition) {
        smoothScrollToPosition(adapterPosition);
      }
    }
  }

  @Override
  public void onClick(View v) {

  }

  @Override
  public void smoothScrollToPosition(int position) {
    final RecyclerView.SmoothScroller smoothScroller = new CenterSmoothScroller(getContext());
    smoothScroller.setTargetPosition(position);
    post(new Runnable() {
      @Override
      public void run() {
        layoutManager.startSmoothScroll(smoothScroller);
      }
    });
  }

  public void setDate(LocalDate date) {
    Day firstDay = adapter.getItem(OFFSET);
    final int offset = (int) ChronoUnit.DAYS.between(firstDay.getDate(), date);
    selectItem(offset + OFFSET);
    ((LinearLayoutManager) getLayoutManager()).scrollToPositionWithOffset(offset, 0);
  }

  private static class CenterSmoothScroller extends LinearSmoothScroller {

    CenterSmoothScroller(Context context) {
      super(context);
    }

    @Override
    public int calculateDtToFit(int viewStart, int viewEnd, int boxStart, int boxEnd, int snapPreference) {
      return (boxStart + (boxEnd - boxStart) / 2) - (viewStart + (viewEnd - viewStart) / 2);
    }
  }
}

File: horizontal-picker/src/main/java/com/github/jhonnyx2012/horizontalpicker/HorizontalPickerRecyclerView.kt
package com.github.jhonnyx2012.horizontalpicker

import android.content.Context
import android.util.AttributeSet
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.LinearSmoothScroller
import androidx.recyclerview.widget.LinearSnapHelper
import androidx.recyclerview.widget.RecyclerView
import org.joda.time.DateTime
import org.joda.time.Days

/**
 * Created by Jhonny Barrios on 22/02/2017.
 */
class HorizontalPickerRecyclerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : RecyclerView(context, attrs, defStyle), OnItemClickedListener, View.OnClickListener {

    private var adapter: HorizontalPickerAdapter? = null
    private var lastPosition = 0
    private var layoutManager: LinearLayoutManager? = null
    private var itemWidth = 0f
    private var listener: HorizontalPickerListener? = null
    private var offset = 0

    fun initialize(
        daysToPlus: Int,
        initialOffset: Int,
        mBackgroundColor: Int,
        mDateSelectedColor: Int,
        mDateSelectedTextColor: Int,
        mTodayDateTextColor: Int,
        mTodayDateBackgroundColor: Int,
        mDayOfWeekTextColor: Int,
        mUnselectedDayTextColor: Int
    ) {
        offset = initialOffset
        layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        setLayoutManager(layoutManager)
        post {
            itemWidth = measuredWidth / 7f
            adapter = HorizontalPickerAdapter(
                itemWidth.toInt(),
                this@HorizontalPickerRecyclerView,
                daysToPlus,
                initialOffset,
                mBackgroundColor,
                mDateSelectedColor,
                mDateSelectedTextColor,
                mTodayDateTextColor,
                mTodayDateBackgroundColor,
                mDayOfWeekTextColor,
                mUnselectedDayTextColor
            )
            setAdapter(adapter)
            val snapHelper = LinearSnapHelper()
            snapHelper.attachToRecyclerView(this@HorizontalPickerRecyclerView)
            removeOnScrollListener(onScrollListener)
            addOnScrollListener(onScrollListener)
        }
    }

    private val onScrollListener: OnScrollListener = object : OnScrollListener() {
        override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
            super.onScrollStateChanged(recyclerView, newState)
            when (newState) {
                SCROLL_STATE_IDLE -> {
                    listener?.onStopDraggingPicker()
                    val position = (computeHorizontalScrollOffset() / itemWidth + 3.5).toInt()
                    if (position != -1 && position != lastPosition) {
                        selectItem(true, position)
                        selectItem(false, lastPosition)
                        lastPosition = position
                    }
                }
                SCROLL_STATE_DRAGGING -> listener?.onDraggingPicker()
            }
        }
    }

    private fun selectItem(isSelected: Boolean, position: Int) {
        adapter?.run {
            getItem(position).isSelected = isSelected
            notifyItemChanged(position)
            if (isSelected) {
                listener?.onDateSelected(getItem(position))
            }
        }
    }

    fun setListener(listener: HorizontalPickerListener) {
        this.listener = listener
    }

    override fun onClickView(view: View, adapterPosition: Int) {
        if (adapterPosition != lastPosition) {
            selectItem(true, adapterPosition)
            selectItem(false, lastPosition)
            lastPosition = adapterPosition
        }
    }

    override fun onClick(view: View) {
        setDate(DateTime())
    }

    override fun smoothScrollToPosition(position: Int) {
        val smoothScroller: SmoothScroller = CenterSmoothScroller(context)
        smoothScroller.targetPosition = position
        post { layoutManager?.startSmoothScroll(smoothScroller) }
    }

    fun setDate(date: DateTime) {
        val today = DateTime().withTime(0, 0, 0, 0)
        val difference =
            Days.daysBetween(date, today).days * if (date.year < today.millis) -1 else 1
        smoothScrollToPosition(offset + difference)
    }

    private class CenterSmoothScroller(context: Context) : LinearSmoothScroller(context) {
        override fun calculateDtToFit(
            viewStart: Int,
            viewEnd: Int,
            boxStart: Int,
            boxEnd: Int,
            snapPreference: Int
        ): Int {
            return boxStart + (boxEnd - boxStart) / 2 - (viewStart + (viewEnd - viewStart) / 2)
        }
    }
}